File: golang/src/p3/main.go
package main

import "fmt"

// Largest prime factor of 600851475143
func main() {
	factors := getFactors(600851475143)

	var max uint64 = 0
	for _, factor := range factors {
		if factor > max {
			max = factor
		}
	}

	fmt.Println(max)
}

func getFactors(n uint64) []uint64 {
	result := make([]uint64, 0)

	for i := uint64(2); i < n; i++ {
		if n%i == 0 {
			result = append(result, i)
			n /= i
			i = 1
		}
	}

	if n != 1 {
		result = append(result, n)
	}

	return result
}

File: cpp/src/p3/main.cc
#include <iostream>
#include <vector>

std::vector<uint64_t> getFactors(uint64_t n) {
  std::vector<uint64_t> result;
  for (uint64_t i = 2u; i < n; ++i) {
    if ((n%i) == 0) {
      result.push_back(i);
      n /= i;
      i = 1;
    }
  }

  if (n != 1) {
    result.push_back(n);
  }
  return result;
}

// Largest prime factor of 600851475143
int main() {
  const auto factors = getFactors(600851475143u);
  uint64_t max = 0u;
  for (const auto& factor : factors) {
    if (factor > max) {
      max = factor;
    }
  }
  std::cout << max << "\n";
}
